Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,404 @@
+                self.seenFirstResponse = true
+            }
+        case .upgraderReady(let upgrade):
+            if case .end = responsePart {
+                // This is the end of the first response, and we can upgrade. Time to kick it off.
+                self.seenFirstResponse = true
+                upgrade()
+            }
+        case .upgradeFailed:
+            // We were reentrantly called while delivering the response head. We can just pass this through.
+            context.fireChannelRead(data)
+        case .upgradeComplete:
+            //Upgrade has completed but we have not seen a whole response and still got reentrantly called.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.receivedResponseAfterUpgradeCompleted)
+        case .requestRequired:
+            //We are receiving an upgrade response and we have not requested the upgrade.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.receivedResponseBeforeRequestSent)
+        }
+    }
+    
+    private func firstResponseHeadReceived(context: ChannelHandlerContext, responsePart: HTTPClientResponsePart) {
+        
+        // We should decide if we're can upgrade based on the first response header: if we aren't upgrading,
+        // by the time the body comes in we should be out of the pipeline. That means that if we don't think we're
+        // upgrading, the only thing we should see is a response head. Anything else in an error.
+        guard case .head(let response) = responsePart else {
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.invalidHTTPOrdering)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Assess whether the upgrade response has accepted our upgrade request.
+        guard case .switchingProtocols = response.status else {
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Ok, we have a HTTP response. Check if it's an upgrade confirmation.
+        // If it's not, we want to pass it on and remove ourselves from the channel pipeline.
+        let acceptedProtocols = response.headers[canonicalForm: "upgrade"].map(String.init)
+        
+        guard acceptedProtocols.count > 0 else {
+            // There is no upgrade protocols returned dspite status being 'switching protocols'.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.responseProtocolNotFound)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        if let callback = self.handleUpgrade(context: context,
+                                             upgradeResponse: response,
+                                             receivedProtocols: acceptedProtocols) {
+            self.gotUpgrader(upgrader: callback)
+        } else {
+            self.notUpgrading(context: context, data: responsePart)
```

Comment:
There's a subtle correctness problem around `notUpgrading` here. `notUpgrading` correctly transitions the internal state of this class into `upgradeFailed` before it makes outcalls, ensuring that any re-entrant operation will be correctly processed. However, it leaves the `fireErrorCaught` calls to the caller, which means they occur before the internal state has changed. Any re-entrant operation occurring from there will likely lead to a crash.

`notUpgrading` should probably accept an `Error?` which it can unwrap and fire through the pipeline after the state change has occurred. If it does that then we can actually rewrite a lot of this code to use `throws` and have a single top-level failure block, plus the one early return where the response is not switching protocols.